### ISSUE TITLE
put context caching functionality under context.cache

### ISSUE THREAD

**Issue by billba** [Microsoft/botbuilder-dotnet#262]
(adapted from https://github.com/Microsoft/botbuilder-js/issues/66#issuecomment-368210787)

Let's put the context cache functionality under the name `context.Cache` (.net, where it would presumably be a `Dictionary`), and `context.cache` (JS, where it would be a `Map`): 

## C#

```csharp
if (context.Cache.ContainsKey("myKey")) {
    value = context.Cache["myKey"};
    context.Cache.Remove("myKey");
}
```

## JavaScript

```ts
if (context.cache.has("mykey")) {
    value = context.cache.get("myKey");
    context.cache.delete("myKey");
}
```

This strikes a better balance between common naming across SDKs and language-specific idiom. 

In the current JS implementation, `context` just exposes `get/has/set` - functions, which by themselves don't telegraph much about their purpose. This puts them behind a meaningful identifier.

Most bot developers will never used the cache directly, and this reduces the surface area of `context` (the list that pops up under intellisense).

Finally, by exposing the underlying object directly, the technical debt of this feature goes down.

**Comment by drub0y** [comment]
Hmmm.... if I'm understanding correctly what this is, I'm not a huge fan of "cache" as the name. What `IBotContext::Get/Set` currently do is store "services" that are available to the scope of the request. So what do you think about the term "services" here?

To me "cache" implies something about state and, even though some sort of caching tech (e.g. Redis) might be used as an implementation for state management, the framework currently refers to this concept as "state". So adding another term like "cache" in here might be a little confusing.

FWIW, I found the `Get/Set` handing right off `IBotContext` wasn't exactly the clearest design either, so I had actually started some prototyping that moves the `Get/Set` off `IBotContext` directly and instead added a `Services` property that has a dedicated interface allowing for adding and getting services to the `IBotContext`. Effectively the same thing, but moving it to a property/interface that is dedicated to that set of functionality "felt" better, more discoverable and decoupled the responsibility from `IBotContext` itself.

**Comment by billba** [comment]
That's not correct, which is exactly why we need to get the naming right.

Services have different caching needs. If you have a RegEx-based recognizer, it can cache over the lifetime of the app instance. A LUIS-based recognizer can also have a long-lived cache, but it should get invalidated if the model changes.

A very common need is to cache over the length of a turn. A state manager does this. And so the question is, where can it cache its values? That's where `context.cache` comes in. So for example (apologies for the JS):

```ts
const conversationState = new ConversationState(new MyDBStorage(...));

bot
    .use(async (context, next) => {
        const state = await conversationState.get(context);
        // write to state
    })
    .onReceive(context => {
        const state = await conversationState.get(context);
        // read from state
    });
```

If you looked inside `conversationState.get(context)` it might look something like:

```
const key = `microsoft/bill/stateManager/${context.channelId}/${context.conversation.id}`;
if (context.cache.has(key)) {
    return context.cache.get(key);
} else {
    const state = await storageProvider.get(`${context.channelId}/${context.conversation.id}`);
    context.cache.set(key, state);
    return state;
}
```

... and then later on, at the end of the turn, middleware would call `conversationState.set(context)`, which might look something like:

```
const key = `microsoft/bill/stateManager/${context.channelId}/${context.conversation.id}`;
if (context.cache.has(key)) {
    const state = context.cache.get(key);
    await storageProvider.set(`${context.channelId}/${context.conversation.id}`, state);
}
```

To summarize:
* Some services cache over the lifetime of a turn
* That's what `context.cache` is for
* Customers of that service never directly interact with the cache

Make sense?

**Comment by drub0y** [comment]
I spoke with @billba 1:1 about this today and have a clearer understanding what he's looking for. Maybe I'm the only one who didn't get this, but let me try to restate it here to make sure we're all on the same page and please correct me if I'm still not getting this right somehow Bill: 

Effectively it's a "cache as a service" that is tied to the lifetime of the `IBotContext` (e.g. a "per turn" cache) that all the players in the request pipeline can utilize rather than having to roll their own approach to that kind of caching.

Now the state manager itself obviously needs to do this in order to provide read-through semantics to the underlying store so that the first read goes through to the store and subsequent reads just pull straight from the state cache, but that seems like the canonical example. Are there any other, non-contrived, examples for this that prove that it's truly worth making this a first class service provided by the bot framework and exposed as a first class concept via the `IBotContext` itself? It seems small enough, but every extra API introduced to the surface area comes with a cost.

If it's not intrinsic to the `IBotContext`, but still "in the box" as a service, say `IBotContextCacheService`, that was able to be registered via middleware and elegantly accessed off of the `IBotContext` via a friendly extension method (e.g. `.GetCache()`), it would mean that some piece of middleware that depends on that service would have to document that dependency and the bot owner would need to know to enable it in its configuration. Plus that middleware would need to throw a runtime exception when/if it didn't find `IBotContextCacheService`. This is technically how request features work in [ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/request-features#middleware-and-request-features), so there is some precedent out there for this approach.

**Comment by Stevenic** [comment]
For what it's worth I originally brought up with @tomlm that using the term `service` in the signature of `get()/set()` seemed too restrictive given that the context can cache anything with services just being one of the things you can use it to cache.

There are actually lots of non-contrived examples where there's value in using the context as a cache. State is one but others include:

- LuisRecognizer needs to cache the results of it's call to LUIS on the context object as your bot is likely to need to inspect the returned intents from multiple places within your code. The current pattern we're promoting for this is that you just call `LuisRecognizer.recognize()` anytime you need to know the intents and it the actual request will only be made once per turn.
- The BatchOutput middleware is currently using the context object to cache a shared array of responses that can be added to from anywhere within your code.  Again, the pattern we're promoting is that you can just create a new BatchOutput object wrapped around the context object and use fluent methods to append responses to this shared array.
- Pretty much any call to a cognitive service will have a similar pattern to the LuisRecognizer and will want to cache it's result on first access.

Where we started was that middleware would simply tack its results onto the context object in a public way that they could be directly consumed off the context object. This basically worked but had a few issues around typing and potential namespace collisions when you'd add multiple recognizers for instance.  So for `m2` we've moved to a model where instead of directly extending the context object we have you go through accessors exposed off the middleware, or other static methods, that have you pass in the context object to work with the output of a given piece of middleware.  This approach has the added benefit that now a lot of things like the LuisRecognizer don't need to be used as middleware at all. But while we may have changed the access pattern for a lot of these things it doesn't change the fact that we still want to cache the output of most of these (especially the service calls) for the lifetime of a single turn. And given that the this is the exact lifetime of the context object it just makes sense to cache these things off the context object itself.

**Comment by billba** [comment]
Actually I think LUIS and a lot of other services will have a longer cache lifetime than the turn.

**Comment by cleemullins** [comment]
How would that work in a bot that is deployed to an arbitrary number of servers?

We don’t support conversation affinity at this time.

________________________________

**Comment by billba** [comment]
I don't understand the relevance of scale-out. I'm saying that it makes sense to cache the result of a LUIS model for a given utterance for a long time, because it will be the same result until the model changes.

**Comment by billba** [comment]
There were several instances from this week's hack where people tried to use `context.set`/.`get` for the "wrong" thing. The more descriptively we can name this, e.g. `context.serviceCache`, the more likely it won't be abused.

I also think we could go further and explore even more verbosity. My favorite method name in the world (doesn't everyone have a favorite method name?) is React's `dangerouslySetInnerHTML` because it's wildly descriptive: you can set the HTML but you should think twice and probably Google to see what's so dangerous. So maybe something like `context.cacheReallyJustForServices`. It sounds like a joke, but if it forces people to look it up then its accomplished its goals.

**Comment by billba** [comment]
Another idea - force keys to have a minimum length and maybe some slashes, again to emphasize that they are namespaced.

**Comment by cleemullins** [comment]
The current implementation of State has been pretty well settled at this time. We may revisit this post Build.

**Comment by billba** [comment]
I don't quite understand your comment - this DCR was adopted (we eventually settled on `context.service`/`context.Service`.
